src/backend/concurrency/eager_write_txn_manager.cpp

```diff
@@ -0,0 +1,672 @@
+#include "backend/common/logger.h"
+#include "backend/storage/data_table.h"
+#include "backend/storage/tile_group.h"
+#include "backend/storage/tile_group_header.h"
+
+namespace peloton {
+namespace concurrency {
+
+thread_local EagerWriteTxnContext *current_txn_ctx;
+
+EagerWriteTxnManager &EagerWriteTxnManager::GetInstance() {
+  static EagerWriteTxnManager txn_manager;
+  return txn_manager;
+}
+
+// Visibility check
+// check whether a tuple is visible to current transaction.
+// in this protocol, we require that a transaction cannot see other
+// transaction's local copy.
+VisibilityType EagerWriteTxnManager::IsVisible(
+    const storage::TileGroupHeader *const tile_group_header,
+    const oid_t &tuple_id) {
+  txn_id_t tuple_txn_id = tile_group_header->GetTransactionId(tuple_id);
+  cid_t tuple_begin_cid = tile_group_header->GetBeginCommitId(tuple_id);
+  cid_t tuple_end_cid = tile_group_header->GetEndCommitId(tuple_id);
+
+  bool own = (current_txn->GetTransactionId() == tuple_txn_id);
+  bool activated = (current_txn->GetBeginCommitId() >= tuple_begin_cid);
+  bool invalidated = (current_txn->GetBeginCommitId() >= tuple_end_cid);
+
+  if (tuple_txn_id == INVALID_TXN_ID) {
+    // the tuple is not available.
+    if (activated && !invalidated) {
+      // deleted tuple
+      return VISIBILITY_DELETED;
+    } else {
+      // aborted tuple
+      return VISIBILITY_INVISIBLE;
+    }
+  }
+
+  // there are exactly two versions that can be owned by a transaction.
+  // unless it is an insertion.
+  if (own == true) {
+    if (tuple_begin_cid == MAX_CID && tuple_end_cid != INVALID_CID) {
+      assert(tuple_end_cid == MAX_CID);
+      // the only version that is visible is the newly inserted/updated one.
+      return VISIBILITY_OK;
+    } else if (tuple_end_cid == INVALID_CID) {
+      // tuple being deleted by current txn
+      return VISIBILITY_DELETED;
+    } else {
+      // old version of the tuple that is being updated by current txn
+      return VISIBILITY_INVISIBLE;
+    }
+  } else {
+    if (tuple_txn_id != INITIAL_TXN_ID) {
+      // if the tuple is owned by other transactions.
+      if (tuple_begin_cid == MAX_CID) {
+        // in this protocol, we do not allow cascading abort. so never read an
+        // uncommitted version.
+        return VISIBILITY_INVISIBLE;
+      } else {
+        // the older version may be visible.
+        if (activated && !invalidated) {
+          return VISIBILITY_OK;
+        } else {
+          return VISIBILITY_INVISIBLE;
+        }
+      }
+    } else {
+      // if the tuple is not owned by any transaction.
+      if (activated && !invalidated) {
+        return VISIBILITY_OK;
+      } else {
+        return VISIBILITY_INVISIBLE;
+      }
+    }
+  }
+}
+
+void EagerWriteTxnManager::RemoveReader() {
+  LOG_INFO("release all reader lock");
+
+  // Remove from the read list of accessed tuples
+  auto txn = current_txn;
+  auto &rw_set = txn->GetRWSet();
```

Comment:
@ravi-teja-mullapudi first suggested, and I agree, that `auto` is being a little bit overused here. It can make the code really hard to read because you need to guess all types or look up what the method returns to figure it out. Perhaps limit the auto usage to iterators or where the type is very verbose.